pypose/module/dynamics.py

```diff
@@ -0,0 +1,342 @@
+        so that the model is consistent with, e.g., the LTI model and the iterative LQR
+        solver.
+    '''
+
+    def __init__(self, time=False):
+        super().__init__()
+        self.jacargs = {'vectorize':True, 'strategy':'reverse-mode'}
+        if time:
+            self.register_buffer('t',torch.zeros(1))
+            self.register_forward_hook(self.forward_hook)
+
+    def forward_hook(self, module, inputs, outputs):
+        self.state, self.input = inputs
+        self.t.add_(1)
+
+    def forward(self, state, input, time=None):
+        r'''
+        Defines the computation performed at every call.
+
+        Should be overridden by all subclasses.
+
+        Note:
+            The :obj:`forward` method implicitly increments the time step via :obj:`forward_hook`.
+            As a result, in the implementations of :obj:`state_transition` and :obj:`observation`,
+            one does not need to include the time as the input; instead, one can directly access
+            the current time via the data member :obj:`self.t`.
+        '''
+        return self.state_transition(state, input), self.observation(state, input)
+
+    def state_transition(self, state, input, time=None):
+        r'''
+        Args:
+            state : Tensor
+                    The state of the dynamical system
+            input : Tensor
+                    The input to the dynamical system
+
+        Returns:
+            Tensor: The state of the system at next time step
+
+        Note:
+            The users need to define this method and can access the current time via the data
+            member :obj:`self.t`.
+        '''
+        raise NotImplementedError("The users need to define their own state transition method")
+
+    def observation(self, state, input, time=None):
+        r'''
+        Args:
+            state : Tensor
+                    The state of the dynamical system
+            input : Tensor
+                    The input to the dynamical system
+
+        Returns:
+            Tensor: The observation of the system at the current step
+
+        Note:
+            The users need to define this method and can access the current time via the data
+            member :obj:`self.t`.
+        '''
+        raise NotImplementedError("The users need to define their own observation method")
+
+    def reset(self,t=0):
+        self.t.fill_(t)
+
+    def set_linearization_point(self, state, input, t):
+        r'''
+        Function to set the point about which the system is to be linearized.
+
+        Parameters
+        ----------
+        state : Tensor
+                The state of the dynamical system
+        input : Tensor
+                The input to the dynamical system
+        t     : Tensor
+                The time step of the dynamical system
+
+        Returns
+        ----------
+        None
+        '''
+        self.state, self.input, self.t = state, input, t
```

Comment:
This [forward_hook line](https://github.com/pypose/pypose/blob/82602d76014c9291533d9811c2aea3f6143f3fcc/pypose/module/dynamics.py#L83) was originally for this purpose. Maybe we should add an argument to `forward` such as `time=None`, so that time can also be stored.

Reply:
I am cosidering removing `forward_hook` or this `set_linearization_point`, as they are repeated.

Reply:
`forward_hook` indeed could be removed, as it might cause changes in the data members that are not visible to the users.  For `set_linearization_point`, my concern is whether one would like to involve `forward` where the back-propagation is not needed.  How about making a function called `set_states`, that does what's in 154, and then the `forward` can call this function before the time stepping, while if one just wants to do linearization, s/he can just call `set_states`.

Reply:
How about we change to something like this:

```
class System(nn.Module):
    def __init__(self):
        super().__init__()
        self.jacargs = {'vectorize':True, 'strategy':'reverse-mode'}

    def forward(self, state, input, time=None):
        self.state, self.input = input, state
        state = self.state_transition(state, input, time)
        output = self.observation(state, input, time)
        if time is not None:
            return state, output, time + 1
        else:
            return state, output

    def state_transition(self, state, input, time=None):
        raise NotImplementedError("The users need to define their own state transition method")

    def observation(self, state, input, time=None):
        raise NotImplementedError("The users need to define their own observation method")

    def reset(self, t=0):
        self.t.fill_(t)

    @property
    def A(self):
        return self.A_(self.state, self.input)

    @A.setter
    def A(self, A):
        self._A = A

    def A_(self, state, input):
        if hasattr(self, '_A'):
            return self._A
        func = lambda x: self.state_transition(x, input)
        return jacobian(func, state, **self.jacargs)
```
In this case, the user can use `A` directly with `mysystem.A`, which in default linearize the last system point.
Or if a linearization point is needed then user can call `mysystem.A_(state=state, input=input)`. 
Then we don't need to set states or inputs.

Reply:
I think it is a good idea to have `A_` as you defined, though `time` should be in the input as well.
As for the `forward`, perhaps we can replace `reset` with `step_counter` to encapsulate the changes in time?
```
    def forward(self, state, input, time=None):
        self.state, self.input = input, state
        self.step_counter(time)
        new_state = self.state_transition(self.state, self.input, self.time)
        output = self.observation(self.state, self.input, self.time)
        return new_state, output, self.time
    def step_counter(self, time=None):
        self.time = self.time+1 if time is None else time
        return self.time
```

Reply:
1. when `time` is not passed to `forward`, it shouldn't return `time`. But in this example, it always returns `time`, so that the user has to always use `x, y, _ = system(x, u)` instead of `x, y = system(x, u)`.
2. the `state_transition` function should use current `time`, not `next time`? but `self.step_counter` is called before `state_transition`.
3. if it is called `step_counter`, then we should use `step` instead of `time`?
4. The original `reset` is to let the user restart a system in their program, not for the counter.

Reply:
After the Zoom discussion, I tried to implement the changes.  But what we discussed still might not make sense...

Consider this example:
```
class Floquet(System):
    def __init__(self):
        super(Floquet, self).__init__(time=True)

    def state_transition(self, state, input, t=None):
        cc = torch.cos(2*math.pi*self._t/10)
        ss = torch.sin(2*math.pi*self._t/10)
        A = torch.tensor([
            [1., cc/10],
            [cc/10, 1.]])
        B = torch.tensor([
            [ss],
            [1.]])
        return (A.matmul(state.reshape(-1,1)) + B * input).squeeze()
```

By default we let the user use `self._t` as the time variable in `state_transition`/`observation`.  This actually effectively makes the argument `t` useless.  When linearizing, the code would end up doing
```
        func = lambda x: self.state_transition(x, self.input, self.time)
        return jacobian(func, self.state, **self.jacargs)
```
but the `self.time` would not be used at all, which defeats the purpose of using `set_ref_point`.

I might be missing something...

Reply:
One possibility is to -
```
    def forward(self, state, input):
        self.state, self.input = state, input
        return self.state_transition(state, input, self._t), self.observation(state, input, self._t)
    def state_transition(self, state, input, t):
        blah
```
so that `forward` always uses `self._t`, which is the system time.  But the user has to use the argument `t` of `state_transition` as the time variable, instead of `self._t`.

BTW, it seems now `time=False` in `__init__` is meaningless - the user just defines `state_transition` and whether the system is time-varying just depends on whether they use `t`.

Reply:
Implemented and tested what I suggested - see my latest commit.